File: jit/internal/mem.go
/*
 * onejit - JIT compiler in Go
 *
 * Copyright (C) 2018-2020-2020 Massimiliano Ghilardi
 *
 *     This Source Code Form is subject to the terms of the Mozilla Public
 *     License, v. 2.0. If a copy of the MPL was not distributed with this
 *     file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 *
 * mem.go
 *
 *  Created on Dec 20, 2020
 *      Author Massimiliano Ghilardi
 */

package internal

// memory address.
type Mem struct {
	kind Kind
	ro   bool // readonly?
	addr Expr
}

func (m Mem) IsAssignable() bool {
	return !m.ro
}

func MakeMem(kind Kind, addr Expr) Mem {
	KindMustBeUintptrOrPtr("Mem", addr.Kind())
	return Mem{
		kind: kind,
		ro:   false,
		addr: addr,
	}
}

// return a read-only view of a subset of memory contents
// note: memory does not become immutable - it can still be modified
// through the original struct
// note: the current implementation assumes little-endian!
func (m Mem) ReadOnly(subset Kind) Mem {
	KindMustBeSubset("Mem", subset, m.kind)
	return Mem{kind: subset, ro: true, addr: m.addr}
}

func (m Mem) Addr() Expr {
	return m.addr
}

// implement Expr interface
func (m Mem) Class() Class {
	return MEM
}

func (m Mem) Op() Op {
	return VAR
}

func (m Mem) Kind() Kind {
	return m.kind
}

func (m Mem) expr() {}

func (m Mem) RegId() RegId {
	return NoRegId
}

func (m Mem) Size() Size {
	return m.kind.Size()
}

func (m Mem) Children() int {
	return 1
}

func (m Mem) Child(i int) Node {
	if i == 0 {
		return m.addr
	}
	return BadIndex(i, 1)
}

func (m Mem) IsConst() bool {
	// memory access cannot be a compile-time constant
	return false
}

func (m Mem) IsPure() bool {
	// reading a memory address has the same side effects
	// as computing the address to be read
	return m.addr.IsPure()
}

File: onejit/ir/mem.cpp
/*
 * onejit - JIT compiler in C++
 *
 * Copyright (C) 2018-2021 Massimiliano Ghilardi
 *
 *     This Source Code Form is subject to the terms of the Mozilla Public
 *     License, v. 2.0. If a copy of the MPL was not distributed with this
 *     file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * mem.cpp
 *
 *  Created on Jan 20, 2021
 *      Author Massimiliano Ghilardi
 */

#include <onejit/code.hpp>
#include <onejit/func.hpp>
#include <onejit/ir/mem.hpp>

namespace onejit {
namespace ir {

Node Mem::create(Func &func, Kind kind, OpN op, Exprs args) noexcept {
  return Base::create_indirect(func,                  //
                               Header{MEM, kind, op}, //
                               Nodes{args.data(), args.size()});
}

const Fmt &Mem::format(const Fmt &fmt, Syntax syntax, size_t depth) const {
  fmt << '(' << op();
  if (syntax == Syntax::CapstoneCompat) {
    fmt << kind().bitsize();
  } else {
    fmt << '_' << kind().stringsuffix();
  }
  for (size_t i = 0, n = children(); i < n; i++) {
    Node node = child(i);
    if (node) {
      fmt << ' ';
      node.format(fmt, syntax, depth + 1);
    }
  }
  return fmt << ')';
}

} // namespace ir
} // namespace onejit
